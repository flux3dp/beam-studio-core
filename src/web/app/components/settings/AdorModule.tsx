import React, { useCallback, useMemo } from 'react';

import Controls from 'app/components/settings/Control';
import constant, { WorkAreaModel } from 'app/actions/beambox/constant';
import LayerModule from 'app/constants/layer-module/layer-modules';
import moduleOffsets from 'app/constants/layer-module/module-offsets';
import SelectControl from 'app/components/settings/SelectControl';
import UnitInput from 'app/widgets/Unit-Input-v2';
import useI18n from 'helpers/useI18n';
import { OptionValues } from 'app/constants/enums';

interface Props {
  defaultUnit: string;
  selectedModel: WorkAreaModel;
  printAdvancedModeOptions: { value: OptionValues; label: string; selected: boolean }[];
  updateBeamboxPreferenceChange: (item_key: string, newVal: any) => void;
  currentModuleOffsets: { [m: number]: [number, number] };
}

const AdorModule = ({
  defaultUnit,
  selectedModel,
  printAdvancedModeOptions,
  updateBeamboxPreferenceChange,
  currentModuleOffsets,
}: Props): JSX.Element => {
  const lang = useI18n();
  const getModuleOffset = useCallback(
    (module: LayerModule) => currentModuleOffsets[module] || moduleOffsets[module],
    [currentModuleOffsets]
  );

  const editValue = useCallback((module: LayerModule, axis: 'x' | 'y', value: number) => {
    const index = axis === 'x' ? 0 : 1;
    const moduleOffset = [...getModuleOffset(module)];
    console.log(module, currentModuleOffsets, moduleOffsets, moduleOffset);
    moduleOffset[index] = value;
    updateBeamboxPreferenceChange('module-offsets', { ...currentModuleOffsets, [module]: moduleOffset });
  }, [currentModuleOffsets, getModuleOffset, updateBeamboxPreferenceChange]);
  const workareaWidth = useMemo(() => constant.dimension.getWidth(selectedModel) / 10, [selectedModel]);
  const workareaHeight = useMemo(() => constant.dimension.getHeight(selectedModel) / 10, [selectedModel]);
  return (
    <>
      <div className="subtitle">{lang.settings.groups.ador_modules}</div>
      <SelectControl
        label="Printing Advanced Mode"
        id="print-advanced-mode"
        options={printAdvancedModeOptions}
        onChange={(e) => updateBeamboxPreferenceChange('print-advanced-mode', e.target.value)}
      />
      <Controls label={`${lang.layer_module.laser_10w_diode} Offset`}>
        <span className="font2" style={{ marginRight: '10px', lineHeight: '32px' }}>X</span>
        <UnitInput
          id="10w-laser-x-offset"
          unit={defaultUnit === 'inches' ? 'in' : 'mm'}
          min={-workareaWidth}
          max={workareaWidth}
          defaultValue={getModuleOffset(LayerModule.LASER_10W_DIODE)[0]}
          getValue={(val) => editValue(LayerModule.LASER_10W_DIODE, 'x', val)}
          forceUsePropsUnit
          className={{ half: true }}
        />
        <span className="font2" style={{ marginRight: '10px', lineHeight: '32px' }}>Y</span>
        <UnitInput
          id="10w-laser-y-offset"
          unit={defaultUnit === 'inches' ? 'in' : 'mm'}
          min={-workareaHeight}
          max={workareaHeight}
          defaultValue={getModuleOffset(LayerModule.LASER_10W_DIODE)[1]}
          getValue={(val) => editValue(LayerModule.LASER_10W_DIODE, 'y', val)}
          forceUsePropsUnit
          className={{ half: true }}
        />
      </Controls>
      <Controls label={`${lang.layer_module.printing} Offset`}>
        <span className="font2" style={{ marginRight: '10px', lineHeight: '32px' }}>X</span>
        <UnitInput
          id="printer-x-offset"
          unit={defaultUnit === 'inches' ? 'in' : 'mm'}
          min={-workareaWidth}
          max={workareaWidth}
          defaultValue={getModuleOffset(LayerModule.PRINTER)[0]}
          getValue={(val) => editValue(LayerModule.PRINTER, 'x', val)}
          forceUsePropsUnit
          className={{ half: true }}
        />
        <span className="font2" style={{ marginRight: '10px', lineHeight: '32px' }}>Y</span>
        <UnitInput
          id="printer-y-offset"
          unit={defaultUnit === 'inches' ? 'in' : 'mm'}
          min={-workareaHeight}
          max={workareaHeight}
          defaultValue={getModuleOffset(LayerModule.PRINTER)[1]}
          getValue={(val) => editValue(LayerModule.PRINTER, 'y', val)}
          forceUsePropsUnit
          className={{ half: true }}
        />
      </Controls>
      <Controls label={`${lang.layer_module.laser_2w_infrared} Offset`}>
        <span className="font2" style={{ marginRight: '10px', lineHeight: '32px' }}>X</span>
        <UnitInput
          id="2w-ir-laser-x-offset"
          unit={defaultUnit === 'inches' ? 'in' : 'mm'}
          min={-workareaWidth}
          max={workareaWidth}
          defaultValue={getModuleOffset(LayerModule.LASER_1064)[0]}
          getValue={(val) => editValue(LayerModule.LASER_1064, 'x', val)}
          forceUsePropsUnit
          className={{ half: true }}
        />
        <span className="font2" style={{ marginRight: '10px', lineHeight: '32px' }}>Y</span>
        <UnitInput
          id="2w-ir-laser-y-offset"
          unit={defaultUnit === 'inches' ? 'in' : 'mm'}
          min={-workareaHeight}
          max={workareaHeight}
          defaultValue={getModuleOffset(LayerModule.LASER_1064)[1]}
          getValue={(val) => editValue(LayerModule.LASER_1064, 'y', val)}
          forceUsePropsUnit
          className={{ half: true }}
        />
      </Controls>
    </>
  );
};

export default AdorModule;
