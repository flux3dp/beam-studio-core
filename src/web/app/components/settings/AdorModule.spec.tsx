import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import LayerModule from 'app/constants/layer-module/layer-modules';
import moduleOffsets from 'app/constants/layer-module/module-offsets';

import AdorModule from './AdorModule';

jest.mock('helpers/useI18n', () => () => ({
  settings: {
    groups: {
      ador_modules: 'ador_modules',
    },
  },
}));
jest.mock('app/components/settings/Control', () => (
  { label, children }: { label: string; children: React.ReactNode }
) => (
  <div>
    <div>Mock Control{label}</div>
    {children}
  </div>
));
jest.mock('app/widgets/Unit-Input-v2', () => (
  { id, defaultValue, getValue }: {
    id: string;
    defaultValue: number;
    getValue: (val: number) => void;
  }
) => (
  <div>
    <input
      id={id}
      data-testid={id}
      type="number"
      value={defaultValue}
      onChange={(e) => getValue(parseFloat(e.target.value))}
    />
  </div>
));

const mockUpdateBeamboxPreferenceChange = jest.fn();
describe('test AdorModule', () => {
  it('should render correctly', () => {
    const mockInitValue: { [m: number]: [number, number] } = {
      [LayerModule.LASER_10W_DIODE]: [10, 10],
    };

    const { container } = render(<AdorModule
      defaultUnit="mm"
      selectedModel="ado1"
      updateBeamboxPreferenceChange={mockUpdateBeamboxPreferenceChange}
      currentModuleOffsets={mockInitValue}
    />);
    expect(container).toMatchSnapshot();
  });

  test('edit value', () => {
    const mockInitValue: { [m: number]: [number, number] } = {
      [LayerModule.LASER_10W_DIODE]: [10, 10],
    };

    const { rerender, getByTestId } = render(<AdorModule
      defaultUnit="mm"
      selectedModel="ado1"
      updateBeamboxPreferenceChange={mockUpdateBeamboxPreferenceChange}
      currentModuleOffsets={mockInitValue}
    />);
    let input = getByTestId('10w-laser-y-offset') as HTMLInputElement;
    fireEvent.change(input, { target: { value: '20' } });
    expect(mockUpdateBeamboxPreferenceChange).toBeCalledTimes(1);
    expect(mockUpdateBeamboxPreferenceChange).toHaveBeenLastCalledWith('module-offsets', {
      [LayerModule.LASER_10W_DIODE]: [10, 20],
    });
    mockInitValue[LayerModule.LASER_10W_DIODE] = [10, 20];
    rerender(<AdorModule
      defaultUnit="mm"
      selectedModel="ado1"
      updateBeamboxPreferenceChange={mockUpdateBeamboxPreferenceChange}
      currentModuleOffsets={mockInitValue}
    />);
    input = getByTestId('printer-x-offset') as HTMLInputElement;
    fireEvent.change(input, { target: { value: '30' } });
    expect(mockUpdateBeamboxPreferenceChange).toBeCalledTimes(2);
    expect(mockUpdateBeamboxPreferenceChange).toHaveBeenLastCalledWith('module-offsets', {
      [LayerModule.LASER_10W_DIODE]: [10, 20],
      [LayerModule.PRINTER]: [30, moduleOffsets[LayerModule.PRINTER][1]],
    });
  });
});
