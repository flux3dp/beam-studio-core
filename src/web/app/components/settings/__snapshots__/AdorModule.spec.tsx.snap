// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test AdorModule should render correctly 1`] = `
<div>
  <div
    class="subtitle"
  >
    ador_modules
  </div>
  <div>
    <div>
      Mock Control
      laser_10w_diode Offset
    </div>
    <span
      class="font2"
      style="margin-right: 10px; line-height: 32px;"
    >
      X
    </span>
    <div>
      <input
        data-testid="10w-laser-x-offset"
        id="10w-laser-x-offset"
        type="number"
        value="10"
      />
    </div>
    <span
      class="font2"
      style="margin-right: 10px; line-height: 32px;"
    >
      Y
    </span>
    <div>
      <input
        data-testid="10w-laser-y-offset"
        id="10w-laser-y-offset"
        type="number"
        value="10"
      />
    </div>
  </div>
  <div>
    <div>
      Mock Control
      printing Offset
    </div>
    <span
      class="font2"
      style="margin-right: 10px; line-height: 32px;"
    >
      X
    </span>
    <div>
      <input
        data-testid="printer-x-offset"
        id="printer-x-offset"
        type="number"
        value="0"
      />
    </div>
    <span
      class="font2"
      style="margin-right: 10px; line-height: 32px;"
    >
      Y
    </span>
    <div>
      <input
        data-testid="printer-y-offset"
        id="printer-y-offset"
        type="number"
        value="-13.37"
      />
    </div>
  </div>
  <div>
    <div>
      Mock Control
      laser_2w_infrared Offset
    </div>
    <span
      class="font2"
      style="margin-right: 10px; line-height: 32px;"
    >
      X
    </span>
    <div>
      <input
        data-testid="2w-ir-laser-x-offset"
        id="2w-ir-laser-x-offset"
        type="number"
        value="0"
      />
    </div>
    <span
      class="font2"
      style="margin-right: 10px; line-height: 32px;"
    >
      Y
    </span>
    <div>
      <input
        data-testid="2w-ir-laser-y-offset"
        id="2w-ir-laser-y-offset"
        type="number"
        value="26.95"
      />
    </div>
  </div>
</div>
`;
